middleware/gzip/response_filter.go

```diff
@@ -62,11 +63,15 @@ func (r *ResponseFilterWriter) WriteHeader(code int) {
 	} else {
 		r.ResponseWriter.WriteHeader(code)
 	}
+	r.statusCodeWritten = true
 }
 
 // Write wraps underlying Write method and compresses if filters
 // are satisfied
 func (r *ResponseFilterWriter) Write(b []byte) (int, error) {
+	if !r.statusCodeWritten {
```

Comment:
Yeah its not. You can only write synchronously except you want a disjointed output.